src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs

```diff
@@ -167,30 +179,40 @@ internal static void RecycleSocketAsyncEventArgs(SocketAsyncEventArgs args)
                 }
             }
         }
-        /// <summary>
-        /// Releases all resources owned by the connection
-        /// </summary>
-        protected virtual void Dispose(bool disposing)
+
+        public async Task DisposeAsync()
         {
-            if (disposing)
+            if (!_disposed)
             {
+                _disposed = true;
                 _stopping = true;
                 _output.Reader.CancelPendingRead();
 
-                Task.WaitAll(_sendTask, _receiveTask);
+                await Task.WhenAll(_sendTask, _receiveTask);
```

Comment:
Alternatively, what's the point of `_sendTask` and `_receiveTask` they don't really do anything? Change the ctor to
```
_socketTask = Task.WhenAll(ReceiveFromSocketAndPushToWriterAsync(), ReadFromReaderAndWriteToSocketAsync());
```
Just store the socket task? Then that await becomes
```
await _socketTask;
```
You could even expose that task and find if the socket had errored etc